### ISSUE TITLE
Ignore buffer type casting in ignored modules in FSDP

### ISSUE THREAD

**Issue by zhaojuanmao** [pytorch/pytorch#97791]
### 🐛 Describe the bug

We should ignore buffer type casting in ignored modules during runtime and state dict APIs in FSDP

### Versions

trunk

cc @mrshenli @pritamdamania87 @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by awgu** [comment]
Related: https://github.com/pytorch/pytorch/issues/77141 item 3

**Comment by awgu** [comment]
Assign to @zhaojuanmao to hold for ramp up task

**Comment by awgu** [comment]
**Step 1: Save the Ignored Buffers**
In `FullyShardedDataParallel.__init__()`, FSDP initializes the ignored module state via `_init_ignored_module_state()`:
https://github.com/pytorch/pytorch/blob/2ef6ffdfa10fca1a2ba0e2cd6a6d032c1ca8b696/torch/distributed/fsdp/fully_sharded_data_parallel.py#L368
https://github.com/pytorch/pytorch/blob/2ef6ffdfa10fca1a2ba0e2cd6a6d032c1ca8b696/torch/distributed/fsdp/_init_utils.py#L237-L258

For this task, you can add an `_ignored_buffers` attribute that saves the buffers from ignored modules and assign it to `state._ignored_buffers`.

**Step 2: Change the Lazy Init Buffer Cast**
FSDP's _lazy initialization_ refers to some delayed initialization that happens typically upon the beginning of the 1st forward pass. Lazy initialization marks the point at which the model should be _finalized_, most notably the `FullyShardedDataParallel` wrapping. Without the lazy initialization, we may incorrectly set one `FullyShardedDataParallel` instance as the root when doing manual `FullyShardedDataParallel` wrapping. See [`_lazy_init()`](https://github.com/pytorch/pytorch/blob/2ef6ffdfa10fca1a2ba0e2cd6a6d032c1ca8b696/torch/distributed/fsdp/_runtime_utils.py#L145) for the implementation.

Part of `_lazy_init()` is casting the buffers to their computation dtypes:
https://github.com/pytorch/pytorch/blob/2ef6ffdfa10fca1a2ba0e2cd6a6d032c1ca8b696/torch/distributed/fsdp/_runtime_utils.py#L171-L172
For this task, we want to exclude ignored buffers from this casting.

**Step 3: Change State Dict Buffer Cast**
For this step, you may want to do some independent investigation to validate and see if we could simplify our existing implementation.

By default, `state_dict()` calls with `keep_vars=False`. This means that both parameters and buffers are `detach()`ed:
https://github.com/pytorch/pytorch/blob/8289120ef025c3ec9ee89ef2dc972e26743e852e/torch/nn/modules/module.py#L1724-L1729
`detach()`-ing creates a new `torch.Tensor` variable that shares the same storage (i.e. data). The current FSDP buffer casting happens "in-place", meaning using `.data`.
https://github.com/pytorch/pytorch/blob/2ef6ffdfa10fca1a2ba0e2cd6a6d032c1ca8b696/torch/distributed/fsdp/_runtime_utils.py#L1375-L1379
This `.data = ...` means that the `Tensor` variable is preserved, but its storage changes.

However, with `keep_vars=False` (and the resulting `detach()`), the upcasting of buffers to their original dtypes in the post-hook _does not_ affect the original buffer variables even though the cast happens "in-place". The cast is happening "in-place" on a different `Tensor` variable due to the `detach()`.
https://github.com/pytorch/pytorch/blob/8289120ef025c3ec9ee89ef2dc972e26743e852e/torch/distributed/fsdp/_state_dict_utils.py#L237-L250

We should skip any ignored buffers in this casting so that FSDP does not impose any semantics on ignored buffers. (You might expect that casting the ignored buffer to its original dtype would just be a no-op, but what if the user chooses to change the ignored buffer in-place as a part of the user's program? It is better to just ignore it completely.)

**Step 4: Add Unit Tests**
We should add some unit tests to `test/distributed/fsdp/test_fsdp_state_dict.py` to test this new behavior.
